cmd/charm/charmcmd/logout.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the GPLv3, see LICENCE file for details.
+
+package charmcmd
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/juju/cmd"
+	"github.com/juju/persistent-cookiejar"
+	"golang.org/x/net/publicsuffix"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/macaroon.v1"
+)
+
+type logoutCommand struct {
+	cmd.CommandBase
+}
+
+var logoutDoc = `
+The logout command removes all security credentials for charm store.
+
+   charm logout
+`
+
+func (c *logoutCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "logout",
+		Purpose: "logout from charm store",
+		Doc:     logoutDoc,
+	}
+}
+
+func (c *logoutCommand) Run(ctxt *cmd.Context) error {
+	// Delete any Ubuntu SSO token.
+	if err := os.Remove(ussoTokenPath()); err != nil && !os.IsNotExist(err) {
+		return errgo.New("cannot remove Ubuntu SSO token")
+	}
+
+	// Delete any login cookies.
+	jar, err := cookiejar.New(&cookiejar.Options{
+		PublicSuffixList: publicsuffix.List,
+		Filename:         cookiejar.DefaultCookieFile(),
+	})
+	if err != nil {
+		return errgo.New("cannot load the cookie jar")
+	}
+	defer jar.Save()
+	u, err := url.Parse(serverURL())
+	if err != nil {
+		// If we can't parse the charmstore URL then we can't be
+		// logged in to it.
+		return nil
+	}
+	seen := map[string]bool{
+		u.String(): true,
+	}
+	urls := []*url.URL{u}
+	for i := 0; i < len(urls); i++ {
+		for _, cookie := range jar.AllCookies() {
+			// All login related cookies will be called macaroon-something.
+			if !strings.HasPrefix(cookie.Name, "macaroon-") {
+				continue
+			}
+			if !cookieForURL(cookie, urls[i]) {
+				continue
+			}
+			ms, err := decodeMacaroonSlice(cookie.Value)
+			if err != nil {
+				continue
+			}
+			// Check for any third party caveats in the macaroon for
+			// which we might also need to delete cookies. By
+			// convention these will be found in ms[0].
```

Comment:
There's no particular reason they should only be in the first macaroon. Discharge macaroons can have third party caveats too, even if we don't use that functionality yet (we might do in some cases in fact).